### PR TITLE
When a county runs out of ballots, we can only terminate single county audits

### PR DESCRIPTION
**Pull request by sduckett**
If County A and County B were both participating in some multi-county `Contest One` and County B had a second single-county `Contest Two`, the simple act of running out of ballots to audit in County B (low voter turnout?) would try to end all of County B's audits. That's fine for `Contest Two`, but if we end `Contest One`, then we'll never be able to achieve a risk limit and complete as hoped.

You can see this for yourself if you use `test/e-state-county/audit.bash`. I've fiddled with it so it's a little cleaner in the UI. That'll run an audit for each of the four contests. Two contests, `Prop 1` and `Prop 2` include all three counties, one contest, `Prop 3`, occurs in Alamosa and Arapaho, and the last, `Prop 4`, is only in Alamosa.

### CODE REVIEW

File: server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/SignOffAuditRound.java [hunk 1]
```diff
@@ -184,19 +184,9 @@ public String endpointBody(final Request request,
...
 
-        final AuditBoardDashboardASM asm = ASMUtilities.asmFor(
-            AuditBoardDashboardASM.class,
-            String.valueOf(cdb.id()));
-
-        if (null != asm && asm.currentState() == ROUND_IN_PROGRESS) {
-          ASMUtilities.step(ROUND_COMPLETE_EVENT,
-                            AuditBoardDashboardASM.class,
-                            String.valueOf(cdb.id()));
-        }
```
Comment (gws) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/SignOffAuditRound.java]: if nothing else comes of this, thank you for taking out my trash :recycle:

Comment (sduckett) [reply server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/SignOffAuditRound.java]: I couldn't get anything to finish taking this bit out, so it's back.

File: server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/SignOffAuditRound.java [hunk 2]
```diff
@@ -208,35 +209,30 @@ public String endpointBody(final Request request,
...
-                + " cdb.auditedSampleCount()=%d,"
-                + " cdb.ballotsAudited()=%d]",
-                cdb.county().name(),
-                cdb.estimatedSamplesToAudit(),
-                cdb.auditedSampleCount(),
-                cdb.ballotsAudited()));
             my_event.set(RISK_LIMIT_ACHIEVED_EVENT);
-            cdb.endAudits();
-            audit_complete = true;
+            final List<ComparisonAudit> terminated = cdb.endSingleCountyAudits();
```
Comment (sduckett) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/SignOffAuditRound.java]: In this case, we'd be terminating single county audits for opportunistic benefits only.

File: server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/SignOffAuditRound.java [hunk 3]
```diff
@@ -208,35 +209,30 @@ public String endpointBody(final Request request,
...
+            LOGGER.debug(String.format("[signoff: all targeted audits finished in %s County."
+                                       + " Terminated these audits: %s]",
+                                       cdb.county().name(), terminated));
+            auditComplete = true;
           } else if (cdb.cvrsImported() <= cdb.ballotsAudited()) {
-            LOGGER.debug("[signoff: there are no more ballots in the county]");
-            my_event.set(BALLOTS_EXHAUSTED_EVENT);
-            cdb.endAudits();
-            audit_complete = true;
+            final List<ComparisonAudit> terminated = cdb.endSingleCountyAudits();
```
Comment (sduckett) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/SignOffAuditRound.java]: In this case, we'd be terminating targeted and opportunistic single county audits.

File: server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/SignOffAuditRound.java [hunk 4]
```diff
@@ -184,19 +185,19 @@ public String endpointBody(final Request request,
...
 
-        final AuditBoardDashboardASM asm = ASMUtilities.asmFor(
-            AuditBoardDashboardASM.class,
-            String.valueOf(cdb.id()));
+        final AuditBoardDashboardASM asm =
+          ASMUtilities.asmFor(AuditBoardDashboardASM.class,
+                              String.valueOf(cdb.id()));
 
-        if (null != asm && asm.currentState() == ROUND_IN_PROGRESS) {
+       if (null != asm && asm.currentState() == ROUND_IN_PROGRESS) {
```
Comment (gws) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/SignOffAuditRound.java]: lost a space

File: server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/SignOffAuditRound.java (hunk 3 above)
Comment (gws) [reply server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/SignOffAuditRound.java]: this and your other note might make good inline comments

File: test/e-state-county/audit.bash [hunk 5]
```diff
@@ -17,17 +17,15 @@ mkdir -p $LOGDIR
     ./main.py reset
 
     echo -e "\nStart 'Define audit' step and upload county audit data\n"
-    ./main.py dos_init
+    ./main.py dos_init -s 12341234123412341234 -r 0.05
```
Comment (nealmcb) [inline test/e-state-county/audit.bash]: Can you add a comment (perhaps at the top of the file) on the expected results given this particular seed and risk limit? E.g. perhaps these changes are necessary for this to be a good test case for this PR (assuming that is true)?

**Pull request merged**